examples/step-92/step-92.cc

```diff
@@ -0,0 +1,1091 @@
+
+    std::string mesh_filename, cells_solution_filename;
+
+  private:
+    int default_physical_ix = -1;
+
+    std::list<Material_tuple>          materials_list;
+    std::list<BoundaryPotential_tuple> boundary_values_list;
+
+  public:
+    unsigned long        N_physicals = 0;
+    material_definition *p_physicals = NULL;
+
+    unsigned long       N_boundary_values = 0;
+    boundary_potential *p_boundary_values = NULL;
+  };
+
+
+  template <int dim>
+  ProblemParameters<dim>::ProblemParameters()
+    : ParameterAcceptor("Problem")
+  {
+    // mu_vacuum      = 4.0e-7 * dealii::numbers::PI; // [ V s / (A m) ]
+    epsilon_vacuum = 8.8541878176204199e-12; // [ A s / (V m) ]
+
+
+    mesh_filename = "";
+    add_parameter("mesh filename", mesh_filename, "mesh file name");
+    cells_solution_filename = "";
+    add_parameter("cells solution filename",
+                  cells_solution_filename,
+                  "filename for exporting cells solution");
+
+    add_parameter(
+      "materials",
+      materials_list,
+      "list of material definitions: \"name : ID : eps_r : kappa\"");
+    add_parameter("potentials",
+                  boundary_values_list,
+                  "list of potential definitions: \"name : ID : epot [V]\"");
+  }
+
+  template <int dim>
+  void ProblemParameters<dim>::finalize()
+  {
+    N_physicals = materials_list.size() + 1;
+
+    p_physicals = new material_definition[N_physicals];
+
+    int ix              = 0;
+    default_physical_ix = -1;
+    for (const auto &item : materials_list)
+      {
+        std::string  name  = std::get<0>(item);
+        unsigned int id    = std::get<1>(item);
+        double       eps_r = std::get<2>(item);
+        double       kappa = std::get<3>(item);
+
+        p_physicals[ix].name = name;
+
+        p_physicals[ix].material_id = id;
+        if (id == 0)
+          default_physical_ix = ix;
+
+        p_physicals[ix].epsilon = eps_r * epsilon_vacuum;
+        p_physicals[ix].kappa   = kappa;
+
+        ix++;
+      }
+
+    N_physicals = ix;
+
+    if (default_physical_ix < 0)
+      {
+        N_physicals++;
+        p_physicals[ix].name = "default";
+
+        default_physical_ix         = ix;
+        p_physicals[ix].material_id = 0;
+
+        p_physicals[ix].epsilon = epsilon_vacuum;
+        p_physicals[ix].kappa   = 0.0;
+      }
+
+    N_boundary_values = boundary_values_list.size();
+    p_boundary_values = new boundary_potential[N_boundary_values];
+
+    unsigned int        ipot = 0;
+    boundary_potential *t_p_opdata;
+    for (const auto &item : boundary_values_list)
+      {
+        std::string name      = std::get<0>(item);
+        auto        id        = std::get<1>(item);
+        double      potential = std::get<2>(item);
+
+        t_p_opdata              = &p_boundary_values[ipot];
+        t_p_opdata->material_id = id;
+        t_p_opdata->name        = name;
+        t_p_opdata->potential   = potential;
+
+        ipot++;
+      }
+  }
```

Comment:
I'm sure that you planned to do it, but it would be worth sketching out what these operations in the call to `finalize()` are doing.